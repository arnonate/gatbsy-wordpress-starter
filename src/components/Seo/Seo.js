import React from "react"
import PropTypes from "prop-types"
import { Helmet } from "react-helmet"
import { useStaticQuery, graphql } from "gatsby"

function Seo({ description, meta, title }) {
  const { site } = useStaticQuery(
    graphql`
      query {
        site {
          siteMetadata {
            title
            description
            author
          }
        }
      }
    `
  )

  const metaDescription = description || site.siteMetadata.description

  return (
    <Helmet
      title={title}
      titleTemplate={`%s | ${site.siteMetadata.title}`}
      meta={[
        {
          name: `description`,
          content: metaDescription,
        },
        {
          property: `og:title`,
          content: title,
        },
        {
          property: `og:description`,
          content: metaDescription,
        },
        {
          property: `og:type`,
          content: `website`,
        },
        {
          name: `twitter:card`,
          content: `summary`,
        },
        {
          name: `twitter:creator`,
          content: site.siteMetadata.author,
        },
        {
          name: `twitter:title`,
          content: title,
        },
        {
          name: `twitter:description`,
          content: metaDescription,
        },
      ].concat(meta)}
    />
  )
}

Seo.defaultProps = {
  meta: [],
  description: ``,
}

Seo.propTypes = {
  description: PropTypes.string,
  meta: PropTypes.arrayOf(PropTypes.object),
  title: PropTypes.string.isRequired,
}

export default Seo
