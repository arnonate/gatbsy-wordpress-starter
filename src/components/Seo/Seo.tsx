import React from "react"
import { Helmet } from "react-helmet"
import { useStaticQuery, graphql } from "gatsby"

type SeoProps = {
  description?: string
  meta?: JSX.IntrinsicElements["meta"][]
  title: string
}

type MetaData = {
  site: {
    siteMetadata: {
      title: string
      description: string
      author: string
    }
  }
}

const Seo = ({
  description = "",
  meta = [],
  title,
}: Readonly<SeoProps>): JSX.Element => {
  const { site }: MetaData = useStaticQuery(
    graphql`
      query {
        site {
          siteMetadata {
            title
            description
            author
          }
        }
      }
    `
  )

  const metaDescription: string = description || site.siteMetadata.description

  return (
    <Helmet
      title={title}
      titleTemplate={`%s | ${site.siteMetadata.title}`}
      meta={[
        {
          name: `description`,
          content: metaDescription,
        },
        {
          property: `og:title`,
          content: title,
        },
        {
          property: `og:description`,
          content: metaDescription,
        },
        {
          property: `og:type`,
          content: `website`,
        },
        {
          name: `twitter:card`,
          content: `summary`,
        },
        {
          name: `twitter:creator`,
          content: site.siteMetadata.author,
        },
        {
          name: `twitter:title`,
          content: title,
        },
        {
          name: `twitter:description`,
          content: metaDescription,
        },
      ].concat(meta)}
    />
  )
}

export default Seo
