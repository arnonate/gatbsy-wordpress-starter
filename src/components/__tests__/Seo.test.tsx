import React from "react"
import { render, waitFor } from "@testing-library/react"
import Helmet from "react-helmet"

import { SeoComponent as Seo, DataProps } from "../Seo/Seo"

const data: DataProps = {
  site: {
    siteMetadata: {
      title: "Headless",
      description: "This is the description of Headless.",
      author: "@arnonate",
    },
  },
  image: {
    childImageSharp: {
      fixed: {
        width: 1200,
        height: 650,
        src: "og:image",
        srcSet: "og:image srcSet",
      },
    },
  },
}

describe("Seo", () => {
  it("renders title correctly with default site description and author", async () => {
    render(<Seo title="Headless Page" data={data} />)

    await waitFor(() => {
      const headData = Helmet.peek()
      return expect(headData).toMatchSnapshot()
    })
  })

  it("renders description correctly", async () => {
    render(
      <Seo
        title="Headless Page"
        description="This is a description of Headless Page"
        data={data}
      />
    )

    await waitFor(() => {
      const headData = Helmet.peek()
      return expect(headData).toMatchSnapshot()
    })
  })

  it("renders metadata correctly", async () => {
    render(
      <Seo
        title="Headless Page"
        meta={[
          { name: "meta", content: "content" },
          { property: "meta", content: "content" },
        ]}
        data={data}
      />
    )

    await waitFor(() => {
      const headData = Helmet.peek()
      return expect(headData).toMatchSnapshot()
    })
  })
})
