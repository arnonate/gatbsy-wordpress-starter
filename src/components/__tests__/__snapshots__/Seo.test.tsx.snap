// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Seo renders description correctly 1`] = `
Object {
  "baseTag": Array [],
  "bodyAttributes": Object {},
  "defer": true,
  "encode": true,
  "htmlAttributes": Object {},
  "linkTags": Array [],
  "metaTags": Array [
    Object {
      "content": "This is a description of Headless Page",
      "name": "description",
    },
    Object {
      "content": "Headless Page | Headless",
      "property": "og:title",
    },
    Object {
      "content": "This is a description of Headless Page",
      "property": "og:description",
    },
    Object {
      "content": "website",
      "property": "og:type",
    },
    Object {
      "content": "og:image",
      "property": "og:image",
    },
    Object {
      "content": "summary",
      "name": "twitter:card",
    },
    Object {
      "content": "@arnonate",
      "name": "twitter:creator",
    },
    Object {
      "content": "Headless Page | Headless",
      "name": "twitter:title",
    },
    Object {
      "content": "This is a description of Headless Page",
      "name": "twitter:description",
    },
    Object {
      "content": "og:image",
      "name": "twitter:image",
    },
  ],
  "noscriptTags": Array [],
  "onChangeClientState": [Function],
  "scriptTags": Array [],
  "styleTags": Array [],
  "title": "Headless Page | Headless",
  "titleAttributes": Object {},
}
`;

exports[`Seo renders metadata correctly 1`] = `
Object {
  "baseTag": Array [],
  "bodyAttributes": Object {},
  "defer": true,
  "encode": true,
  "htmlAttributes": Object {},
  "linkTags": Array [],
  "metaTags": Array [
    Object {
      "content": "This is the description of Headless.",
      "name": "description",
    },
    Object {
      "content": "Headless Page | Headless",
      "property": "og:title",
    },
    Object {
      "content": "This is the description of Headless.",
      "property": "og:description",
    },
    Object {
      "content": "website",
      "property": "og:type",
    },
    Object {
      "content": "og:image",
      "property": "og:image",
    },
    Object {
      "content": "summary",
      "name": "twitter:card",
    },
    Object {
      "content": "@arnonate",
      "name": "twitter:creator",
    },
    Object {
      "content": "Headless Page | Headless",
      "name": "twitter:title",
    },
    Object {
      "content": "This is the description of Headless.",
      "name": "twitter:description",
    },
    Object {
      "content": "og:image",
      "name": "twitter:image",
    },
    Object {
      "content": "content",
      "name": "meta",
    },
    Object {
      "content": "content",
      "property": "meta",
    },
  ],
  "noscriptTags": Array [],
  "onChangeClientState": [Function],
  "scriptTags": Array [],
  "styleTags": Array [],
  "title": "Headless Page | Headless",
  "titleAttributes": Object {},
}
`;

exports[`Seo renders title correctly with default site description and author 1`] = `
Object {
  "baseTag": Array [],
  "bodyAttributes": Object {},
  "defer": true,
  "encode": true,
  "htmlAttributes": Object {},
  "linkTags": Array [],
  "metaTags": Array [
    Object {
      "content": "This is the description of Headless.",
      "name": "description",
    },
    Object {
      "content": "Headless Page | Headless",
      "property": "og:title",
    },
    Object {
      "content": "This is the description of Headless.",
      "property": "og:description",
    },
    Object {
      "content": "website",
      "property": "og:type",
    },
    Object {
      "content": "og:image",
      "property": "og:image",
    },
    Object {
      "content": "summary",
      "name": "twitter:card",
    },
    Object {
      "content": "@arnonate",
      "name": "twitter:creator",
    },
    Object {
      "content": "Headless Page | Headless",
      "name": "twitter:title",
    },
    Object {
      "content": "This is the description of Headless.",
      "name": "twitter:description",
    },
    Object {
      "content": "og:image",
      "name": "twitter:image",
    },
  ],
  "noscriptTags": Array [],
  "onChangeClientState": [Function],
  "scriptTags": Array [],
  "styleTags": Array [],
  "title": "Headless Page | Headless",
  "titleAttributes": Object {},
}
`;
